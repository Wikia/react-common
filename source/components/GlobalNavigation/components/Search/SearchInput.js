import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import { useTranslation } from 'react-i18next';

import Icon from '../../../Icon';
import Button from '../../../Button';

import I18nNamespaceContext from '../../context/I18nNamespaceContext';

const SearchInput = React.forwardRef((props, ref) => {
    const {
        model,
        query,
        onSearchActivation,
        onQueryChanged,
        onFocus,
        onBlur,
        onSearchClear,
        onSearchClose,
        onSearchSubmit,
    } = props;
    const placeholderConfig = model['placeholder-active'];
    const [t] = useTranslation(useContext(I18nNamespaceContext));

    return (
        <React.Fragment>
            <input
                className="wds-global-navigation__search-input"
                placeholder={t(placeholderConfig.key, { sitename: placeholderConfig.params.sitename.value })}
                value={query}
                onClick={onSearchActivation}
                ref={ref}
                onChange={onQueryChanged}
                onFocus={onFocus}
                onBlur={onBlur}
                type="search"
            />
            <Button
                className="wds-global-navigation__search-clear"
                type="button"
                onClick={onSearchClear}
                text
            >
                <Icon name="add" small className="wds-global-navigation__search-clear-icon" />
            </Button>
            <Button
                className="wds-global-navigation__search-close"
                type="button"
                onClick={onSearchClose}
                text
            >
                <Icon name="close-tiny" className="wds-global-navigation__search-close-icon" tiny />
            </Button>
            <Button
                className="wds-global-navigation__search-submit"
                type="submit"
                disabled={!query}
                onClick={onSearchSubmit}
                data-tracking-label={model.results['tracking-label']}
            >
                <Icon name="arrow-small" className="wds-global-navigation__search-submit-icon" small />
            </Button>
        </React.Fragment>
    );
});

SearchInput.propTypes = {
    model: PropTypes.shape().isRequired,
    onBlur: PropTypes.func.isRequired,
    onFocus: PropTypes.func.isRequired,
    onQueryChanged: PropTypes.func.isRequired,
    onSearchActivation: PropTypes.func.isRequired,
    onSearchClear: PropTypes.func.isRequired,
    onSearchClose: PropTypes.func.isRequired,
    onSearchSubmit: PropTypes.func.isRequired,
    query: PropTypes.string.isRequired,
};

export default SearchInput;
