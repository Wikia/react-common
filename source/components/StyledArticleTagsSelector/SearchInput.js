import React from 'react';
import styled, { css } from 'styled-components';
import PropTypes from 'prop-types';
import noop from 'lodash/noop';

import IconMagnifyingGlassSmall from '../../icons/IconMagnifyingGlassSmall';
import IconCloseSmall from '../../icons/IconCloseSmall';

import { colorWithAlpha } from '../../utils/css';

import StyledList from '../StyledList';
import HighlightedText from '../HighlightedText';

import TagShape from './TagShape';

const ICON_STYLES = css`
    fill: currentColor;
    height: 16px;
    width: 16px;
`;

const List = styled(StyledList)`
    background-color: ${({ theme }) => theme.color.white};
    border: 1px solid ${({ theme }) => theme.color.fandom_middle_gray};
    overflow: hidden;
    padding: 10px 0;
    position: absolute;
    width: 100%;
    top: 32px;

    @media ${({ theme }) => theme.media.medium_up} {
        top: 24px;
    }

    > li {
        cursor: pointer;
        line-height: 36px;
        margin: 0 10px;
        padding: 0 10px;

        &:hover {
            background-color: ${({ accentColor }) => colorWithAlpha(accentColor, 0.1)};
        }
    }

    mark {
        background-color: transparent;
        font-weight: ${({ theme }) => theme.font_weight.bold};
    }
`;

const Slider = styled.div`
    @media ${({ theme }) => theme.media.small_down} {
        background-color: ${({ theme }) => theme.color.white};
        height: 100vh;
        left: 0;
        position: fixed;
        top: 0;
        width: 100vw;
        z-index: 1;
    }

    @media ${({ theme }) => theme.media.medium_up} {
        flex: 1;
    }
`;

const Wrapper = styled.div`
    align-items: center;
    border-bottom: solid 1px ${({ theme }) => theme.color.black};
    display: flex;
    position: relative;

    @media ${({ theme }) => theme.media.small_only} {
        margin: 13px;
    }
`;

const Input = styled.input`
    background-color: transparent;
    border: none;
    flex-grow: 1;
    font-size: ${({ theme }) => theme.font_size.base};
    line-height: ${({ theme }) => theme.line_height.normal};
    width: 100%;

    @media ${({ theme }) => theme.media.medium_up} {
        font-size: ${({ theme }) => theme.font_size.s};
    }

    &:focus {
        outline: none;
    }
`;

const IconSearch = styled(IconMagnifyingGlassSmall)`
    ${ICON_STYLES}
    margin-right: 6px;
`;

const IconClose = styled(IconCloseSmall)`
    ${ICON_STYLES}

    cursor: pointer;
`;

function SearchInput({
    accentColor,
    className,
    onChange,
    onClose,
    onAddTag,
    communityName,
    list,
    query,
}) {
    const ref = React.createRef();

    return (
        <Slider className={className}>
            <Wrapper>
                <IconSearch />
                <Input
                    placeholder={`Searching ${communityName}`}
                    ref={ref}
                    onChange={onChange}
                    autoFocus
                />
                {list && (
                    <List accentColor={accentColor}>
                        {list.map(({ id, title }) => (
                            <li
                                key={id}
                                onClick={/* istanbul ignore next */ () => onAddTag(id)}
                            >
                                <HighlightedText
                                    highlight={query}
                                    text={title}
                                />
                            </li>
                        ))}
                    </List>
                )}
                <IconClose onClick={onClose} />
            </Wrapper>
        </Slider>
    );
}

SearchInput.propTypes = {
    accentColor: PropTypes.string.isRequired,
    /** Extra class name */
    className: PropTypes.string,
    communityName: PropTypes.string,
    list: PropTypes.arrayOf(TagShape),
    onAddTag: PropTypes.func,
    onChange: PropTypes.func,
    onClose: PropTypes.func,
    query: PropTypes.string,
};

SearchInput.defaultProps = {
    className: '',
    communityName: '',
    list: null,
    onAddTag: noop,
    onChange: noop,
    onClose: noop,
    query: '',
};

export default SearchInput;
