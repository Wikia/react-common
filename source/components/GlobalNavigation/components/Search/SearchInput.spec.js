import React from 'react';
import { shallow } from 'enzyme';
import merge from 'lodash/merge';

import SearchInput from './SearchInput';

const defaultProps = {
    query: 'some query',
    model: {
        type: 'search',
        results: {
            'tracking-label': 'search',
            'param-name': 'query',
            url: '//starwars.wikia.com/wiki/Special:Search',
        },
        'placeholder-inactive': {
            type: 'translatable-text',
            key: 'global-navigation-search-placeholder-inactive',
        },
        'placeholder-active': {
            type: 'translatable-text',
            key: 'global-navigation-search-placeholder-in-wiki',
            params: {
                sitename: {
                    type: 'text',
                    value: 'Wookieepedia',
                },
            },
        },
        suggestions: {
            url: '/mocks/searchSuggestions.json?',
            'param-name': 'query',
            'tracking-label': 'search-suggestion',
        },
    },
    onSearchActivation: () => null,
    onQueryChanged: () => null,
    onFocus: () => null,
    onBlur: () => null,
    onSearchClear: () => null,
    onSearchClose: () => null,
    onSearchSubmit: () => null,
};

function renderComponent(props, options = {}) {
    const computedProps = merge({}, defaultProps, props);

    return shallow(<SearchInput {...computedProps} />, options);
}

test('SearchInput renders', () => {
    const component = renderComponent();

    expect(component).toMatchSnapshot();
});
