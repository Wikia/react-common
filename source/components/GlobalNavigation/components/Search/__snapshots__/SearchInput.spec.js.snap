// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput renders 1`] = `
<Fragment>
  <input
    className="wds-global-navigation__search-input"
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    placeholder="global-navigation-search-placeholder-in-wiki"
    type="search"
    value="some query"
  />
  <Button
    className="wds-global-navigation__search-clear"
    disabled={false}
    fullwidth={false}
    href={null}
    light={false}
    onClick={[Function]}
    secondary={false}
    square={false}
    text={true}
    type="button"
  >
    <Icon
      className="wds-global-navigation__search-clear-icon"
      name="add"
      small={true}
      tiny={false}
    />
  </Button>
  <Button
    className="wds-global-navigation__search-close"
    disabled={false}
    fullwidth={false}
    href={null}
    light={false}
    onClick={[Function]}
    secondary={false}
    square={false}
    text={true}
    type="button"
  >
    <Icon
      className="wds-global-navigation__search-close-icon"
      name="close-tiny"
      small={false}
      tiny={true}
    />
  </Button>
  <Button
    className="wds-global-navigation__search-submit"
    data-tracking-label="search"
    disabled={false}
    fullwidth={false}
    href={null}
    light={false}
    onClick={[Function]}
    secondary={false}
    square={false}
    text={false}
    type="submit"
  >
    <Icon
      className="wds-global-navigation__search-submit-icon"
      name="arrow-small"
      small={true}
      tiny={false}
    />
  </Button>
</Fragment>
`;
